engines/chewy/video/cfo_decoder.cpp

```diff
@@ -0,0 +1,310 @@
+	frameType = _fileStream->readUint16LE();
+
+	switch (frameType) {
+	case FRAME_TYPE:
+		handleFrame();
+		break;
+	case CUSTOM_FRAME_TYPE:
+		handleCustomFrame();
+		break;
+	default:
+		error("CfoDecoder::decodeFrame(): unknown main chunk type (type = 0x%02X)", frameType);
+		break;
+	}
+
+	_curFrame++;
+	_nextFrameStartTime += _frameDelay;
+
+	return _surface;
+}
+
+#define FLI_SETPAL 4
+#define FLI_SS2    7
+#define FLI_BRUN   15
+#define FLI_COPY   16
+#define PSTAMP     18
+
+void CfoDecoder::CfoVideoTrack::handleFrame() {
+	uint16 chunkCount = _fileStream->readUint16LE();
+
+	// Read subchunks
+	for (uint32 i = 0; i < chunkCount; ++i) {
+		uint32 frameSize = _fileStream->readUint32LE();
+		uint16 frameType = _fileStream->readUint16LE();
+		uint8 *data = new uint8[frameSize - 6];
+		_fileStream->read(data, frameSize - 6);
+
+		switch (frameType) {
+		case FLI_SETPAL:
+			unpackPalette(data);
+			_dirtyPalette = true;
+			break;
+		case FLI_SS2:
+			decodeDeltaFLC(data);
+			break;
+		case FLI_BRUN:
+			decodeByteRun(data);
+			break;
+		case FLI_COPY:
+			copyFrame(data);
+			break;
+		case PSTAMP:
+			/* PSTAMP - skip for now */
+			break;
+		default:
+			error("FlicDecoder::decodeNextFrame(): unknown subchunk type (type = 0x%02X)", frameType);
+			break;
+		}
+
+		delete[] data;
+	}
+}
+
+void CfoDecoder::CfoVideoTrack::handleCustomFrame() {
+	uint16 chunkCount = _fileStream->readUint16LE();
+
+	uint16 delay, number, channel, volume, repeat, balance;
+	Audio::AudioStream *stream;
+
+	// Read subchunks
+	for (uint32 i = 0; i < chunkCount; ++i) {
+		uint32 frameSize = _fileStream->readUint32LE();
+		uint16 frameType = _fileStream->readUint16LE();
+
+		switch (frameType) {
+		case kChunkFadeIn:
+			error("Unused chunk kChunkFadeIn found");
+			break;
+		case kChunkFadeOut:
+			// Used in video 0
+			delay = _fileStream->readUint16LE();
+
+			warning("kChunkFadeOut, delay %d", delay);
+			// TODO
+			break;
+		case kChunkLoadMusic:
+			// Used in videos 0, 18, 34, 71
+			warning("kChunkLoadMusic");
+			// TODO
+			_fileStream->skip(frameSize);
+			break;
+		case kChunkLoadRaw:
+			error("Unused chunk kChunkLoadRaw found");
+			break;
+		case kChunkLoadVoc:
+			number = _fileStream->readUint16LE();
+			assert(number < MAX_SOUND_EFFECTS);
+			delete[] _soundEffects[number];
+
+			_soundEffectSize[number] = frameSize - 2;
+			_soundEffects[number] = (byte *)malloc(frameSize - 2);
+			_fileStream->read(_soundEffects[number], frameSize - 2);
+			break;
+		case kChunkPlayMusic:
+			// Used in videos 0, 18, 34, 71
+			warning("kChunkPlayMusic");
+			// TODO
+			_fileStream->skip(frameSize);
+			break;
+		case kChunkPlaySeq:
+			error("Unused chunk kChunkPlaySeq found");
+			break;
+		case kChunkPlayPattern:
+			error("Unused chunk kChunkPlayPattern found");
+			break;
+		case kChunkStopMusic:
+			_mixer->stopHandle(_musicHandle);
+			break;
+		case kChunkWaitMusicEnd:
+			do {
+				g_system->delayMillis(10);
```

Comment:
Also add g_system->updateScreen(), and preferably processEvents(), so the mouse cursor does not get frozen here.